src/credentials/GroupDataProviderImpl.h

```diff
@@ -150,25 +153,38 @@ class GroupDataProviderImpl : public GroupDataProvider
     class GroupKeyContext : public Crypto::SymmetricKeyContext
     {
     public:
-        GroupKeyContext(GroupDataProviderImpl & provider) : mProvider(provider) {}
+        GroupKeyContext(GroupDataProviderImpl & provider, Crypto::SessionKeystore & keystore) :
+            mProvider(provider), mSessionKeystore(keystore)
+        {}
+
+        GroupKeyContext(GroupDataProviderImpl & provider, Crypto::SessionKeystore & keystore,
+                        const Crypto::Aes128KeyByteArray & encryptionKey, uint16_t hash,
+                        const Crypto::Aes128KeyByteArray & privacyKey) :
+            mProvider(provider),
+            mSessionKeystore(keystore)
 
-        GroupKeyContext(GroupDataProviderImpl & provider, const ByteSpan & encryptionKey, uint16_t hash,
-                        const ByteSpan & privacyKey) :
-            mProvider(provider)
         {
-            SetKey(encryptionKey, hash);
-            SetPrivacyKey(privacyKey);
+            Initialize(encryptionKey, hash, privacyKey);
         }
 
-        void SetKey(const ByteSpan & encryptionKey, uint16_t hash)
+        void Initialize(const Crypto::Aes128KeyByteArray & encryptionKey, uint16_t hash,
+                        const Crypto::Aes128KeyByteArray & privacyKey)
         {
+            ReleaseKeys();
             mKeyHash = hash;
-            memcpy(mEncryptionKey, encryptionKey.data(), std::min(encryptionKey.size(), sizeof(mEncryptionKey)));
+            // TODO: Load group keys to the session keystore upon loading from persistent storage
+            //
+            // Group keys should be transformed into a key handle as soon as possible or even
+            // the key storage should be taken over by SessionKeystore interface, but this looks
+            // like more work, so let's use the transitional code below for now.
+            mSessionKeystore.CreateKey(encryptionKey, mEncryptionKey);
+            mSessionKeystore.CreateKey(privacyKey, mPrivacyKey);
         }
 
-        void SetPrivacyKey(const ByteSpan & privacyKey)
+        void ReleaseKeys()
         {
-            memcpy(mPrivacyKey, privacyKey.data(), std::min(privacyKey.size(), sizeof(mPrivacyKey)));
+            mSessionKeystore.DestroyKey(mEncryptionKey);
+            mSessionKeystore.DestroyKey(mPrivacyKey);
```

Comment:
Is this safe to call multiple times? What happens if ReleaseKeys() is called many times in a row (e.g. both explicitly and by destructors somewhere)

Reply:
Yes, by definition `DestroyKey` should deinitialize a handle, and `DestroyKey` on uninitialized handle is a no-op.